Реализовано задание спринта 9

karabasBaRaBaS-1276 opened the pull request:
1. Реализована генерация большого кол-ва чисел в слайсе как в отдном, так и в несколько потоков.
2. Реализовано два варианта поиска максимального значения в слайсе.
3. Написаны Unit тесты

rongo-base64 committed:
Add generateRandomElements

rongo-base64 committed:
Add tests for generateRandomElements

rongo-base64 committed:
Add readme

Commit:
done

Commit:
add new test

dirtymew commented on main.go:
```diff
@@ -2,39 +2,132 @@ package main
+	}
+	workers := runtime.NumCPU() // Количество доступных CPU
+	//workers := CHUNKS
+	if (size < MIN_PARALLEL_SIZE) || (workers == 1) {
```
прикольно - но по заданию не требуется

dirtymew commented on main.go:
```diff
@@ -2,39 +2,132 @@ package main
+
+	wg.Add(workers)
+	for w := 0; w < workers; w++ {
+		go func(worker int) {
```
лучше в горутину передавать готовый слайс

dirtymew replied:
так будет правильно с точки зрения архитектуры

dirtymew commented on main_test.go:
```diff
@@ -1,3 +1,144 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGenerateRandomElements_EmptySlice(t *testing.T) {
```
лучше для таких случаев писать табличные тесты

dirtymew commented on main_test.go:
```diff
@@ -1,3 +1,144 @@
+		{
+			name: "Отрицательное число",
+			args: args{
+				data: []int{-789, -200, -100},
```
такого кейса нет - у нас по условию только положительные числа

dirtymew commented on main_test.go:
```diff
@@ -1,3 +1,144 @@
+			want: 1060,
+		},
+		{
+			name: "Число 0",
```
еще хорошо бы кейсы с nil пустым слайсом и с одним элементом

dirtymew commented on main.go:
```diff
@@ -2,39 +2,132 @@ package main
+			}
+
+			result := data[start]
+			for i := start + 1; i < end; i++ {
```
тут ошибка  - нужно использовать maximum

karabasBaRaBaS-1276 replied:
Да. В задании написано, "Чтобы найти максимальное значение быстрее, можно разделить слайс на восемь частей и в каждом из них найти максимум."
Мне показалось, что необязательно использовать срезы. Насколько помню cрез создаёт новый слайс из существующего слайса с сохранением базового. Мне показалось, что реализация будет медленнее, чем та, что у меня. Да и памяти потребует больше. Переделаю и заодно проверю по скорости :)

dirtymew replied:
@karabasBaRaBaS-1276 тут не в том смысл немного - у нас есть функция maximum 

авторы задания хотели показать как горутины могут быть эффективнее... тут речь идет про batching - поэтому для чистоты эксперимента нужно тут туже функцию использовать

по поводу слайдов - тут не будет ничего по памяти алоцироваться заново - слайс ссылочный тип и при передачи в горутину ничего копироваться не будет

чтобы проверять по скорости нужно бенчмарки писать

dirtymew replied:
*слайсов )

dirtymew commented on main.go:
```diff
@@ -2,39 +2,132 @@ package main
+	wg.Wait()
+
+	result := chunkMaxValues[0]
+	for i := 1; i < len(chunkMaxValues); i++ {
```
тут ошибка  - нужно использовать maximum

dirtymew reviewed (changes requested):
Привет Роман

хорошая работа - но нужно немного поправить

Commit:
Исправление замечаний:
1. Теперь горутины работают со срезом оригинального слайса. Возвращают максимальное значение в канал.
2. Добавлена ещё одна горутина для сбора максимальных значений в отдельный слайс.
3. Параметры, с которыми работают горутины теперь явно передаются.
4. Переиспользование функции maximum
5. В тестах пустой слайс перенесен в табличный тест. Убраны тесты на отрицательные значения.

dirtymew commented on main.go:
```diff
@@ -2,39 +2,134 @@ package main
+
+	// --- запускаем горутину, которая будет из канала вычитывать
+	//     значение и складывать в слайс для окончательной обработки
+	ch := make(chan int, 1) // буфер = 1
```
хмм но вот с каналом будет медленне - хотя как тренировка к теме спринта отлично

karabasBaRaBaS-1276 replied:
Да. Я больше для тренировки. 
Но я локально сравнивал с вариантом "по-старому" (не стал включать в коммит итоговый). 
На удивление разница между старой реализацией (когда нет срезов слайсов и каналов) и текущая не отличалась по скорости +- 2мс даже с 1 млрд. элементов. А вот разница с одним потоком существенная (до 40% быстрее многопоточность)

dirtymew commented on main.go:
```diff
@@ -2,39 +2,134 @@ package main
+
+	var wg0 sync.WaitGroup
+	wg0.Add(1)
+	go func(chunkMaxValues []int, ch <-chan int) {
```
тут не нужна вейт группа - у нас только одна горутина...  - нужно только закрыть канал на 112 строке

и прямо тут в горутине посчитать максимум по chunkMaxValues

это будет логичнее...

dirtymew commented on main_test.go:
```diff
@@ -1,3 +1,133 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := maxChunks(tt.args.data); got != tt.want {
+				t.Errorf("maximum() = %v, want %v", got, tt.want)
```
для проверок используй testify

dirtymew reviewed (approved):
Привет Роман

работа принята
